smriprep/utils/bids.py

```diff
@@ -1,14 +1,157 @@
-"""
+"""Utilities to handle BIDS inputs."""
+from collections import defaultdict
+from pathlib import Path
+from json import loads
+from pkg_resources import resource_filename as pkgrf
+from bids.layout.writing import build_path
+
+
+def predict_derivatives(subject_id, output_spaces, freesurfer):
+    """
+    Generate a list of the files that should be found in the output folder.
+
+    The prediction of outputs serves two purposes:
+
+      * Anticipates the set of outputs that will be generated.
+      * Informs the decision of whether the workflow should be staged or some
+        sections can be skip.
+
+    Parameters
+    ----------
+    subject_id : :obj:`str`
+        A subject id
+    output_spaces : :obj:`list`
+        TemplateFlow identifiers of the requested output spaces
+    freesurfer : :obj:`bool`
+        Whether the ``--fs-no-reconall`` flag was used.
+
+    Examples
+    --------
+    >>> predict_derivatives('01', ['MNI152NLin2009cAsym'], False)
+    ['sub-01/anat/sub-01_desc-brain_mask.nii.gz',
+     'sub-01/anat/sub-01_desc-preproc_T1w.nii.gz',
+     'sub-01/anat/sub-01_dseg.nii.gz',
+     'sub-01/anat/sub-01_from-MNI152NLin2009cAsym_to-T1w_mode-image_xfm.h5',
+     'sub-01/anat/sub-01_from-T1w_to-MNI152NLin2009cAsym_mode-image_xfm.h5',
+     'sub-01/anat/sub-01_label-CSF_probseg.nii.gz',
+     'sub-01/anat/sub-01_label-GM_probseg.nii.gz',
+     'sub-01/anat/sub-01_label-WM_probseg.nii.gz',
+     'sub-01/anat/sub-01_space-MNI152NLin2009cAsym_desc-brain_mask.nii.gz',
+     'sub-01/anat/sub-01_space-MNI152NLin2009cAsym_desc-preproc_T1w.nii.gz',
+     'sub-01/anat/sub-01_space-MNI152NLin2009cAsym_dseg.nii.gz',
+     'sub-01/anat/sub-01_space-MNI152NLin2009cAsym_label-CSF_probseg.nii.gz',
+     'sub-01/anat/sub-01_space-MNI152NLin2009cAsym_label-GM_probseg.nii.gz',
+     'sub-01/anat/sub-01_space-MNI152NLin2009cAsym_label-WM_probseg.nii.gz']
+
+    """
+    spec = loads(Path(pkgrf('smriprep', 'data/io_spec.json')).read_text())
+
+    def _normalize_q(query):
+        query['subject'] = subject_id
+        if 'extension' not in query:
+            query['extension'] = ['nii', 'nii.gz']
+        if 'space' in query and not query['space']:
+            query['space'] = output_spaces
+        if 'from' in query and not query['from']:
+            query['from'] = output_spaces
+        if 'to' in query and not query['to']:
+            query['to'] = output_spaces
+        return query
+
+    queries = [_normalize_q(q) for q in spec['queries']['baseline'].values()]
+    if freesurfer:
+        queries += [_normalize_q(q)
+                    for q in spec['queries']['surfaces'].values()]
+
+    output = []
+    for q in queries:
+        paths = build_path(q, spec['patterns'])
+        if isinstance(paths, str):
+            output.append(paths)
+        elif paths:
+            output += paths
+
+    return sorted(output)
+
+
+def collect_derivatives(derivatives_dir, subject_id, std_spaces, freesurfer,
```

Comment:
perhaps we can include a check to the derivatives_dir's `dataset_description.json` here and compare against the current version of smriprep. this is my current thinking of behavior, but it may be too strict?

- Different major (incompatible)
- Different minor (probably incompatible?)
- Different patch (probably fine)

Reply:
So I would not worry much about this for now. Please note that we are using versioning for two different objectives - API compatibility and Data compatibility.

For the API compatibility, this PR should not change sMRIPrep's API at all - meaning, fMRIPrep should keep working exactly the same after these changes make it into a release. This PR should only add a new feature that fMRIPrep and dMRIPrep might want to use.

Data compatibility - the intent here is to allow users to feed in their own derivatives, so if they want to replace sMRIPrep completely, they might do so. The "data API" here is BIDS derivatives, and those that need be found in the output folder, prescribed in the new `smriprep/data/io_spec.json` file. So, the Data API will probably not change, unless fMRIPrep needs new anatomical results. So, if any, changes will go downstream, never upstream.

None of this I just said addresses the problem of reproducibility, I'm well aware of that. Some possible measures we might want to take are described here https://github.com/poldracklab/smriprep/pull/107#issuecomment-575809877 - but I believe that is out of the scope of this PR.

I agree with @effigies that splitting the smriprep outputs outside the fmriprep output folder would help navigate these issues easier. That would allow us to check on a `dataset_description.json` particular of sMRIPrep (as I think you were suggesting). At this point, only the fMRIPrep's one is being generated and it is being checked. All these changes pertain to fMRIPrep/dMRIPrep proper, so I guess we can use the latter to experiment these measures.